pkg/internal/controller/controller_test.go

```diff
@@ -1014,6 +1103,207 @@ var _ = Describe("controller", func() {
+				}),
+			}
+
+			// Wait for the goroutines to finish before returning to avoid racing with the
+			// assignment in BeforeEach block.
+			var wg sync.WaitGroup
+
+			// Invoked in a goroutine because Warmup will block
+			wg.Add(1)
+			go func() {
+				defer GinkgoRecover()
+				defer wg.Done()
+				Expect(ctrl.Warmup(ctx)).To(Succeed())
+			}()
+
+			cancel()
+			wg.Wait()
+		})
+
+		It("should be called before leader election runnables if warmup is enabled", func() {
+			// This unit test exists to ensure that a warmup enabled controller will actually be
+			// called in the warmup phase before the leader election runnables are started. It
+			// catches regressions in the controller that would not implement warmupRunnable from
+			// pkg/manager.
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			By("Creating a channel to track execution order")
+			runnableExecutionOrderChan := make(chan string, 2)
+			const nonWarmupRunnableName = "nonWarmupRunnable"
+			const warmupRunnableName = "warmupRunnable"
+
+			ctrl.CacheSyncTimeout = time.Second
+			ctrl.startWatches = []source.TypedSource[reconcile.Request]{
+				source.Func(func(ctx context.Context, _ workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+					runnableExecutionOrderChan <- warmupRunnableName
+					return nil
+				}),
+			}
+
+			nonWarmupCtrl := New[reconcile.Request](ControllerOptions[reconcile.Request]{
+				MaxConcurrentReconciles: 1,
+				Do:                      fakeReconcile,
+				NewQueue: func(string, workqueue.TypedRateLimiter[reconcile.Request]) workqueue.TypedRateLimitingInterface[reconcile.Request] {
+					return queue
+				},
+				LogConstructor: func(_ *reconcile.Request) logr.Logger {
+					return log.RuntimeLog.WithName("controller").WithName("test")
+				},
+				CacheSyncTimeout: time.Second,
+				EnableWarmup:     ptr.To(false),
+				LeaderElected:    ptr.To(true),
+			})
+			nonWarmupCtrl.startWatches = []source.TypedSource[reconcile.Request]{
+				source.Func(func(ctx context.Context, _ workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+					runnableExecutionOrderChan <- nonWarmupRunnableName
+					return nil
+				}),
+			}
+
+			By("Creating a manager")
+			testenv = &envtest.Environment{}
+			cfg, err := testenv.Start()
+			Expect(err).NotTo(HaveOccurred())
+			m, err := manager.New(cfg, manager.Options{
+				LeaderElection: false,
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Adding warmup and non-warmup controllers to the manager")
+			Expect(m.Add(ctrl)).To(Succeed())
+			Expect(m.Add(nonWarmupCtrl)).To(Succeed())
+
+			By("Starting the manager")
+			go func() {
+				defer GinkgoRecover()
+				Expect(m.Start(ctx)).To(Succeed())
+			}()
+
+			<-m.Elected()
+			Expect(<-runnableExecutionOrderChan).To(Equal(warmupRunnableName))
+			Expect(<-runnableExecutionOrderChan).To(Equal(nonWarmupRunnableName))
+		})
+
+		It("should not race with Start and only start sources once", func() {
```

Comment:
is the first part true? I mean Start and Warmup are racing (but "only start sources once" is true)

Reply:
By `race`, I meant `should not called a race condition when called concurrently with Start`. Let me reword to make it clear